prototypes/matrix_free_stokes_global_coarsening/matrix_free_stokes_global_coarsening.cc

```diff
@@ -0,0 +1,1738 @@
+  amg_data.smoother_sweeps = mg_data.coarse_solver.smoother_sweeps;
+  amg_data.n_cycles        = mg_data.coarse_solver.n_cycles;
+  amg_data.smoother_type   = mg_data.coarse_solver.smoother_type.c_str();
+
+  precondition_amg.initialize(mg_matrices[min_level]->get_system_matrix(),
+                              amg_data);
+
+  mg_coarse =
+    std::make_unique<MGCoarseGridIterativeSolver<VectorType,
+                                                 SolverGMRES<VectorType>,
+                                                 LevelMatrixType,
+                                                 decltype(precondition_amg)>>(
+      coarse_grid_solver, *mg_matrices[min_level], precondition_amg);
+
+  Multigrid<VectorType> mg(
+    mg_matrix, *mg_coarse, mg_transfer, mg_smoother, mg_smoother);
+
+  PreconditionerType preconditioner(dof, mg, mg_transfer);
+
+  SolverGMRES<VectorType>(solver_control)
+    .solve(fine_matrix, dst, src, preconditioner);
+}
+
+
+template <typename VectorType, typename OperatorType, int dim>
+void
+solve_with_gmg(SolverControl &            solver_control,
+               const OperatorType &       system_matrix,
+               VectorType &               dst,
+               const VectorType &         src,
+               const MultigridParameters &mg_data,
+               const MappingQ<dim>        mapping,
+               const DoFHandler<dim> &    dof_handler,
+               const QGauss<1> &          quadrature)
+{
+  MGLevelObject<DoFHandler<dim>>                     dof_handlers;
+  MGLevelObject<std::unique_ptr<OperatorType>>       operators;
+  MGLevelObject<MGTwoLevelTransfer<dim, VectorType>> transfers;
+
+  std::vector<std::shared_ptr<const Triangulation<dim>>>
+    coarse_grid_triangulations;
+
+  coarse_grid_triangulations =
+    MGTransferGlobalCoarseningTools::create_geometric_coarsening_sequence(
+      dof_handler.get_triangulation());
+
+  const unsigned int n_h_levels = coarse_grid_triangulations.size();
+
+  unsigned int minlevel = 0;
+  unsigned int maxlevel = n_h_levels - 1;
+
+  dof_handlers.resize(minlevel, maxlevel);
+  operators.resize(minlevel, maxlevel);
+  transfers.resize(minlevel, maxlevel);
+
+  for (unsigned int l = minlevel; l <= maxlevel; ++l)
+    {
+      dof_handlers[l].reinit(*coarse_grid_triangulations[l]);
+      dof_handlers[l].distribute_dofs(dof_handler.get_fe());
+    }
+
+  MGLevelObject<AffineConstraints<typename VectorType::value_type>> constraints(
+    minlevel, maxlevel);
+
+  for (unsigned int level = minlevel; level <= maxlevel; ++level)
+    {
+      const auto &dof_handler = dof_handlers[level];
+      auto &      constraint  = constraints[level];
+
+      const IndexSet locally_relevant_dofs =
+        DoFTools::extract_locally_relevant_dofs(dof_handler);
+      constraint.reinit(locally_relevant_dofs);
+
+      DoFTools::make_hanging_node_constraints(dof_handler, constraint);
+      VectorTools::interpolate_boundary_values(mapping,
```

Comment:
I guess in production code this (bcs) would need to be modularized, since it is very much problem specific. (and its duplicated code)